### PR TITLE
Support vision attention for Llama4

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

Add more options to the existing Attention class to support the attention in vision encoder:
* Full attention (non-causal)
* use bias in qkvo projection
* add is_vision flag to attention module for vision specific settings
Future item: we may implement a separate VisionAttention, as the vision dimensions are different from text, which increases the complexity if reuse the current attention module. We are working on the design.

# Tests
Test by comparing outputs with Llama4 VisionAttention module
Screenshot 2025-05-14 at 3 50 43 PM
Currently dot_product attention works as expected, there are some issues with flash attention. Will enable flash attention in future PRs

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/attentions.py [hunk 1]
```diff
@@ -710,7 +711,10 @@ def tpu_flash_attention(
     )
 
     mask_shape = (self.config.max_target_length, self.config.max_target_length)
-    mask = splash_attention_mask.CausalMask(shape=mask_shape)
+    if self.attention_type == AttentionType.FULL:
```
Comment (hengtaoguo) [inline MaxText/layers/attentions.py:714]: Does this mean future flash attention mask will be generated here in the branch?

**Review by hengtaoguo — APPROVED** [review]

File: MaxText/layers/attentions.py [hunk 2]
```diff
@@ -710,7 +711,10 @@ def tpu_flash_attention(
...
+      mask = splash_attention_mask.FullMask(mask_shape)
```
Comment (ZhaoyueCheng) [inline MaxText/layers/attentions.py:715]: since we already have the corresponding full mask generation for splash attention can we try if we can directly enable splash attention for the vision transformer?

**Review by ZhaoyueCheng — APPROVED** [review]

File: MaxText/layers/attentions.py (hunk 1 above)
Comment (aireenmei) [reply MaxText/layers/attentions.py:714]: We have been using this branch to generate mask for flash attention. Before this change, we by default started with a causal mask, and add modifications such as chunks on top of it. After this change, we started with a full mask when attention type is full, and likely won't add more modifications due to the conditions for modifications will be False

File: MaxText/layers/attentions.py (hunk 2 above)
Comment (aireenmei) [reply MaxText/layers/attentions.py:715]: I added Full mask for flash attention here, but still hit other issues, mainly related to the shape not meeting the block_size requirements for flash attention. Will keep working on it.

File: MaxText/tests/check_llama4_layers.py [hunk 3]
```diff
@@ -337,5 +347,222 @@ def __init__(self):
...
+      self,
+      hidden_states: torch.Tensor,
+      freqs_ci: torch.Tensor,
+      attention_mask: Optional[torch.Tensor] = None,
+      past_key_value: Optional[torch.Tensor] = None,
+      **kwargs,
+  ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+    input_shape = hidden_states.shape[:-1]
+    hidden_shape = (*input_shape, -1, self.head_dim)
+    print(hidden_shape)
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply MaxText/tests/check_llama4_layers.py:424]: Good catch. Fixed

File: MaxText/layers/attentions.py (hunk 2 above)
Comment (ZhaoyueCheng) [reply MaxText/layers/attentions.py:715]: yeah makes sense! Thanks Aireen!

File: MaxText/layers/attentions.py [hunk 4]
```diff
@@ -1208,6 +1212,7 @@ class Attention(nn.Module):
   ragged_block_size: int = 256
   use_qk_norm: bool = False
   query_pre_attn_scalar: float | None = None
+  use_bias_in_projection: bool = False
```
Comment (gagika) [inline MaxText/layers/attentions.py:1215]: nit: `use_bias_in_projections` (plural) and please add comment:
looks like it's being applied to many projections (q, k, v) and output projections.

File: MaxText/layers/attentions.py [hunk 5]
```diff
@@ -1394,7 +1403,15 @@ def apply_rotary_embedding(self, inputs: Array, inputs_positions: Array, name: s
 
     rope_type = self.config.rope_type.lower()
     rope_use_scale = self.config.rope_use_scale
-    if self.config.model_name.startswith("llama3.1") or rope_type.startswith("llama3.1"):
+    if self.config.model_name.startswith("llama4") and self.name == "self_attention_vision":
```
Comment (gagika) [inline MaxText/layers/attentions.py:1406]: could you replace `self.name == "self_attention_vision"` with an explicit boolean or enum parameter in the Attention?

e.g. `is_vision_attention`?

Another option is to have VisionAttention as a subclass of Attention that overrides 

class VisionAttention(Attention):
    def apply_rotary_embedding ...

File: MaxText/layers/attentions.py (hunk 4 above)
Comment (aireenmei) [reply MaxText/layers/attentions.py]: fixed

File: MaxText/layers/attentions.py (hunk 5 above)
Comment (aireenmei) [reply MaxText/layers/attentions.py]: I updated to use enum RopeType to replace rope_type to model config. So we can set rope_type at attention level instead of setting one at model level.

Comment (aireenmei) [reply MaxText/layers/attentions.py]: I see you are OOO. Could @RissyRan review this change, also changes in Llama4 and Deepseek?

File: MaxText/configs/models/deepseek2-16b.yml [hunk 6]
```diff
@@ -40,8 +40,6 @@ kv_lora_rank: 512
 qk_nope_head_dim: 128
 qk_rope_head_dim: 64
 v_head_dim: 128
-# RoPE
-rope_type: "yarn"
```
Comment (RissyRan) [inline MaxText/configs/models/deepseek2-16b.yml:44]: I feel this remove the flexibility of rope in models. For instance, how we apply llama3.1 type to a new future deepseek model?

File: MaxText/layers/attentions.py [hunk 7]
```diff
@@ -407,18 +416,18 @@ def generate_attention_mask(
     elif causal_mask is not None:
       output_mask = causal_mask
 
-    is_seq_len_greater_than_chunk_window = False
```
Comment (RissyRan) [inline MaxText/layers/attentions.py:410]: Be careful on this as output_mask is updated in the block `self.attention_type == AttentionType.LOCAL_SLIDING and output_mask is not None`. Worth to double check the logic here.

**Review by RissyRan — COMMENTED** [review]
Thanks Aireen! Just minor comments.

File: MaxText/configs/models/deepseek2-16b.yml (hunk 6 above)
Comment (aireenmei) [reply MaxText/configs/models/deepseek2-16b.yml]: I reverted this change as discussed offline. Thanks for your insights!

File: MaxText/layers/attentions.py (hunk 5 above)
Comment (aireenmei) [reply MaxText/layers/attentions.py]: I discussed with @RissyRan offline, I ended up adding an is_vision flag to the attention module. We may add a visionAttention module in the future (see the future item in PR description). But for now we use this way to get an early version to work.

**Review by RissyRan — APPROVED** [review]

File: MaxText/input_pipeline/_input_pipeline_utils.py
Comment (aireenmei) [inline MaxText/input_pipeline/_input_pipeline_utils.py:1]: This is for solving pylint error

File: MaxText/layers/embeddings.py
Comment (aireenmei) [inline MaxText/layers/embeddings.py:1]: This is for solving presubmit error about API arg counts

File: MaxText/input_pipeline/_input_pipeline_utils.py
Comment (hengtaoguo) [reply MaxText/input_pipeline/_input_pipeline_utils.py:1]: Thanks! There's also another PR noticed the same issue. @gobbleturk  or @SurbhiJainUSC  are working on a fix soon: https://github.com/AI-Hypercomputer/maxtext/pull/1762/files#diff-f4632952887d83be3fc9646878680a77dab4b70af4159c7218e71b860754a55e
